### PR TITLE
feat(toolbox): add end conference

### PR DESCRIPTION
**Pull request by wfleischer**
This PR adds a feature which allows moderators to end the conference for all participants.

It depends on PR https://github.com/jitsi/lib-jitsi-meet/pull/1859.

Screenshot 2022-01-25 at 10 30 09

Screenshot_20220125-104805

### CODE REVIEW

**Comment by shawnchin** [comment]
UX wise, does this mean users now needs 2 clicks to leave a conference? Even for non-moderators? I can also see folks accidentally ending conference instead of leaving which can be rather disruptive.

Could we consider leaving the button's default action being the same as it is now i.e. to leave the conference, but for moderators you have the extra carat icon (like for audio/video) where moderators can explicitly pull up the extra option to end meetings for all?

**Comment by saghul** [comment]
Thanks a lot for this Werner! This feature is in fact in our roadmap. Let me check the status and see if we can meet in the middle.

**Comment by wfleischer** [comment]
> Could we consider leaving the button's default action being the same as it is now i.e. to leave the conference, but for moderators you have the extra carat icon (like for audio/video) where moderators can explicitly pull up the extra option to end meetings for all?

Thank you @shawnchin for your valuable feedback! These are very good thoughts. I do see a value in some kind of confirmation whether to leave the conference (most of *our* users are technically not very skilled). Eventually this could be made a configuration option. I have no strong opinion on this matter. Let's hear what the Jitsi design team recommends. I will then adjust accordingly.

**Comment by wfleischer** [comment]
> Thanks a lot for this Werner! This feature is in fact in our roadmap. Let me check the status and see if we can meet in the middle.

Thank you, @saghul. This is good to know. Whatever this feature will look like. I will be happy to see it being available :smile:

**Comment by wfleischer** [comment]
Marked this as WIP for now. This PR is pending on some UI redesign. No ETA. Sorry.

**Comment by wfleischer** [comment]
Sorry, I did not have cycles recently to work on the UI patch. Fortunately @nbeckindy helped out and provided the web UI part. Here's how it looks like:

File: react/features/base/conference/middleware.any.js
Comment (saghul) [inline react/features/base/conference/middleware.any.js:132]: We might need to trigger the readyToClose event for the iframe API, see below for an example of when we do that.

File: react/features/toolbox/components/native/HangupMenu.js
Comment (saghul) [inline react/features/toolbox/components/native/HangupMenu.js:1]: We are using TS for all new components, can you please adjust? It should be reasonably straightforward now.

File: react/features/toolbox/components/native/HangupMenuButton.js
Comment (saghul) [inline react/features/toolbox/components/native/HangupMenuButton.js:1]: Ditto.

File: react/features/toolbox/components/native/styles.js
Comment (saghul) [inline react/features/toolbox/components/native/styles.js:151]: Please avoid using `schemeColor`, we are phasing that out. Use the design tokens, we now have some genefic Buttons and a destructive flag which sets the right color, for example.

File: react/features/toolbox/components/web/EndConferenceButton.js
Comment (saghul) [inline react/features/toolbox/components/web/EndConferenceButton.js:35]: style: unnecessary newline

File: react/features/toolbox/components/web/EndConferenceButton.js
Comment (saghul) [inline react/features/toolbox/components/web/EndConferenceButton.js:60]: Extract all of these to a constant object, since they don't change. Just leave the props that do change with the redux state here.

File: react/features/toolbox/components/web/LeaveConferenceButton.js
Comment (saghul) [inline react/features/toolbox/components/web/LeaveConferenceButton.js:48]: Since we now have this code in a couple of places how about we create a new action `leaveConference` which encapsulates the behavior?

File: react/features/toolbox/components/web/LeaveConferenceButton.js
Comment (saghul) [inline react/features/toolbox/components/web/LeaveConferenceButton.js:67]: Ditto, extract the constants to outside here.

File: react/features/toolbox/components/web/Toolbox.js
Comment (saghul) [inline react/features/toolbox/components/web/Toolbox.js:25]: put the all caps constants at the top please.

File: resources/prosody-plugins/mod_end_conference.lua
Comment (saghul) [inline resources/prosody-plugins/mod_end_conference.lua:1]: I think it's no longer necessary to split the component modules like this, see the breakout rooms module ;-)

File: resources/prosody-plugins/mod_end_conference_component.lua
Comment (saghul) [inline resources/prosody-plugins/mod_end_conference_component.lua:1]: Add some documentation at the top please.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL. I'll ask some colleagues to look at the CSS parts too.

File: react/features/base/conference/middleware.any.js
Comment (wfleischer) [reply react/features/base/conference/middleware.any.js:132]: Thanks `The meeting has been terminated` is now sent.

File: resources/prosody-plugins/mod_end_conference.lua
Comment (wfleischer) [reply resources/prosody-plugins/mod_end_conference.lua:16]: Migrated code into a single file.

File: react/features/base/color-scheme/defaultScheme.js
Comment (saghul) [inline react/features/base/color-scheme/defaultScheme.js:45]: Please use a token instead. This file will disappear soon. See base/ui/

File: react/features/base/conference/actions.js
Comment (saghul) [inline react/features/base/conference/actions.js:594]: I think this only works if you have joined the conference, so I'd say we should extract the `conference` field from the state. The current conference could be in the joining state.

File: react/features/base/toolbox/components/ToolboxItem.native.js
Comment (saghul) [inline react/features/base/toolbox/components/ToolboxItem.native.js:42]: Can't you use the new `Button` component instead of modifying these?

File: react/features/toolbox/components/native/HangupMenu.js
Comment (saghul) [inline react/features/toolbox/components/native/HangupMenu.js:54]: Why do you need this container? IIRC the bottom sheet implementation should take care of this.

File: react/features/toolbox/components/native/HangupMenu.js
Comment (saghul) [inline react/features/toolbox/components/native/HangupMenu.js:11]: we recently separated sheets from dialogs, there is a hideSheet now.

File: react/features/toolbox/components/native/HangupMenuButton.js
Comment (saghul) [inline react/features/toolbox/components/native/HangupMenuButton.js:43]: it's openSheet now.

File: resources/prosody-plugins/mod_muc_breakout_rooms.lua
Comment (saghul) [inline resources/prosody-plugins/mod_muc_breakout_rooms.lua:224]: What does this change do?

**Review by saghul — COMMENTED** [review]
Left some more feedback.

File: resources/prosody-plugins/mod_muc_breakout_rooms.lua
Comment (wfleischer) [reply resources/prosody-plugins/mod_muc_breakout_rooms.lua:224]: In the case that the main room exists, it tells the participants that they can join the main room, otherwise nil is sent.

File: react/features/toolbox/components/native/HangupMenu.js
Comment (wfleischer) [reply react/features/toolbox/components/native/HangupMenu.js]: Done

File: react/features/toolbox/components/native/HangupMenuButton.js
Comment (wfleischer) [reply react/features/toolbox/components/native/HangupMenuButton.js]: Done

File: react/features/toolbox/components/native/styles.js
Comment (wfleischer) [reply react/features/toolbox/components/native/styles.js]: Got obsolete and has been removed.

File: react/features/toolbox/components/web/EndConferenceButton.js
Comment (wfleischer) [reply react/features/toolbox/components/web/EndConferenceButton.js]: Fixed.

File: react/features/toolbox/components/web/Toolbox.js
Comment (wfleischer) [reply react/features/toolbox/components/web/Toolbox.js]: Fixed.

File: react/features/base/color-scheme/defaultScheme.js
Comment (wfleischer) [reply react/features/base/color-scheme/defaultScheme.js]: Has become obsolete and has been removed.

File: react/features/base/conference/actions.js
Comment (wfleischer) [reply react/features/base/conference/actions.js]: Fixed.

File: react/features/base/toolbox/components/ToolboxItem.native.js
Comment (wfleischer) [reply react/features/base/toolbox/components/ToolboxItem.native.js]: Fixed.

File: react/features/toolbox/components/native/HangupMenu.js
Comment (wfleischer) [reply react/features/toolbox/components/native/HangupMenu.js]: Fixed.

File: react/features/toolbox/components/native/HangupMenu.js
Comment (wfleischer) [reply react/features/toolbox/components/native/HangupMenu.js]: Fixed.

File: react/features/toolbox/components/web/LeaveConferenceButton.js
Comment (wfleischer) [reply react/features/toolbox/components/web/LeaveConferenceButton.js]: Component has been redone. This no longer applies.

**Comment by wfleischer** [comment]
Hi @saghul. This feature is ready for another review. If things are ok, I will be ready to stash-rebase, if you like.

**Comment by saghul** [comment]
> Hi @saghul. This feature is ready for another review. If things are ok, I will be ready to push a squash-rebase, if you like.

Excellent! I'll allocate some time in the next few days!

**Comment by wfleischer** [comment]
@saghul: if https://github.com/jitsi/jitsi-meet/pull/11868 will get merged soon, I would love to make use of the new button component for the end and leave conference web buttons. This way we would not have to modify the OverflowMenuItem and can also remove the button styling.

@robertpin: excellent UI improvement! Thank you.

**Comment by robertpin** [comment]
> @saghul: if #11868 will get merged soon, I would love to make use of the new button component for the end and leave conference web buttons. This way we would not have to modify the OverflowMenuItem and can also remove the button styling.

Thanks for your work @wfleischer . I will update that PR tomorrow and it will hopefully get merged tomorrow

**Comment by wfleischer** [comment]
@saghul: This is ready for another review. I now use the new web Button component. I also force-pushed a squashed rebase, because the commit history had become quite a mess 😄

**Comment by saghul** [comment]
Jenkins please test this please.

File: resources/prosody-plugins/mod_end_conference.lua
Comment (saghul) [inline resources/prosody-plugins/mod_end_conference.lua:67]: I know it's not the same, but don't we usually check for owner here?

File: resources/prosody-plugins/mod_end_conference.lua
Comment (saghul) [inline resources/prosody-plugins/mod_end_conference.lua:74]: should we return true here?

**Review by saghul — CHANGES_REQUESTED** [review]
So sorry it took this long! Did a review round, it's mostly there, just a few small things.

File: react/features/base/conference/middleware.any.js
Comment (wfleischer) [reply react/features/base/conference/middleware.any.js:146]: Without leaving the conference here, after the user clicks end conference, he is left on the conference page with the toolbar and would have to click leave the conference (again) to get to the main screen. This is not obvious. Many users are a bit lost on this "zombie" conference page. Everything looks like they were left alone in the meeting.

Comment (saghul) [reply react/features/base/conference/middleware.any.js:146]: I agree, but as I said we have the same situation with kicking. I'll ask our design team for some ideas here, so let;s drop this part for now and get back to it once I have something. WDYT?

File: react/features/toolbox/components/HangupButton.js
Comment (wfleischer) [reply react/features/toolbox/components/HangupButton.js:48]: Done

File: react/features/toolbox/components/native/HangupMenu.tsx
Comment (wfleischer) [reply react/features/toolbox/components/native/HangupMenu.tsx:42]: Done

File: resources/prosody-plugins/mod_end_conference.lua
Comment (wfleischer) [reply resources/prosody-plugins/mod_end_conference.lua:14]: The end conference reason is checked in _conferenceFailed at react/features/base/conference/middleware.any.js. So I think it has to be a constant.

File: resources/prosody-plugins/mod_end_conference.lua
Comment (wfleischer) [reply resources/prosody-plugins/mod_end_conference.lua:67]: When we check affiliation, we check for owner, while the occupant.role would be either moderator or participant.

File: resources/prosody-plugins/mod_end_conference.lua
Comment (wfleischer) [reply resources/prosody-plugins/mod_end_conference.lua:74]: Good catch!

**Review by wfleischer — COMMENTED** [review]
Thanks, @saghul for your thorough review! Please see the update and comments.

File: react/features/base/conference/actions.js
Comment (wfleischer) [reply react/features/base/conference/actions.js]: Done

File: react/features/base/conference/actions.js
Comment (wfleischer) [reply react/features/base/conference/actions.js]: Makes sense. Done

File: react/features/base/conference/middleware.any.js
Comment (wfleischer) [reply react/features/base/conference/middleware.any.js]: It makes sense to review and unify the end meeting experience in a separate PR. Removed here.

File: react/features/toolbox/components/HangupButton.js
Comment (wfleischer) [reply react/features/toolbox/components/HangupButton.js]: Done

File: react/features/toolbox/components/web/LeaveConferenceButton.tsx
Comment (wfleischer) [reply react/features/toolbox/components/web/LeaveConferenceButton.tsx]: Done

File: react/features/toolbox/components/web/Toolbox.js
Comment (wfleischer) [reply react/features/toolbox/components/web/Toolbox.js]: Done

File: react/features/toolbox/components/native/Toolbox.js
Comment (wfleischer) [reply react/features/toolbox/components/native/Toolbox.js]: Makes sense. Done

**Comment by wfleischer** [comment]
@saghul: pr has been rebased and conflicts resolved. It should now be ready for another review.

**Comment by saghul** [comment]
I think this is good to go, can you rebase one last time pl?

**Comment by damencho** [comment]
Sorry, I have not reviewed and I'm jumping in with a question: If the prosody config is not enabled(serverside disabled) the UI functionality is also hidden right?

**Comment by damencho** [comment]
I guess that's already the case https://github.com/jitsi/lib-jitsi-meet/pull/1859/files#diff-b2f6195536d4c09eac93cd5482c9258cae0dc21129fca78a125fe545594eca30R738

**Comment by wfleischer** [comment]
> I think this is good to go, can you rebase one last time pl?

Thank you, @saghul. Should be ready now. Sorry for the merge-push. I was a bit distracted.

**Review by saghul — APPROVED** [review]

**Comment by saghul** [comment]
Jenkins please test this please.

**Comment by saghul** [comment]
Excellent work @wfleischer ! 👏

**Comment by wfleischer** [comment]
> Excellent work @wfleischer ! 👏

Always a pleasure working with you @saghul and the Jitsi team.

**Comment by wfleischer** [comment]
btw @saghul: Before the next release you would want to merge this related docker-jitsi-meet PR https://github.com/jitsi/docker-jitsi-meet/pull/1208. Other approach would be to disable this feature by default and merge right now and enable it by default after the next release.

**Comment by saghul** [comment]
The master branch in the Docker repo is meant to work with unstable builds, so I merged it and it should be working with tonight's build!

**Comment by wfleischer** [comment]
> The master branch in the Docker repo is meant to work with unstable builds, so I merged it and it should be working with tonight's build!

Right. Hope that you won't be in the need to release a stable-7648-5, though :smile:

**Comment by damencho** [comment]
@saghul @wfleischer 

I enabled end conference on alpha to be able to see it in action, but the tests started failing, so I will disable it.


```
org.openqa.selenium.TimeoutException: 
Expected condition failed: waiting for org.jitsi.meet.test.EndConferenceTest$$Lambda$856/0x00000008402d9840@559cd857 (tried for 5 second(s) with 500 milliseconds interval)
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'ci.jitsi.org', ip: '104.18.18.146', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.0-124-generic', java.version: '11.0.16'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 105.0.5195.52, chrome: {chromedriverVersion: 105.0.5195.19 (b9c217c128c1..., userDataDir: /tmp/.com.google.Chrome.LcYjnL}, goog:chromeOptions: {debuggerAddress: localhost:43027}, javascriptEnabled: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: LINUX, platformName: LINUX, proxy: Proxy(), setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true, webdriver.remote.sessionid: 47a6e131941af6a6cbd96585647...}
Session ID: 47a6e131941af6a6cbd9658564799568
	at org.jitsi.meet.test.EndConferenceTest.hangupCallAndCheck(EndConferenceTest.java:50)
```

We need to adjust the tests to detect whether this is enabled and work in both cases. Its good even to add tests for this functionality if enabled.
Thank you for your contribution.

**Pull request merged**
